command/run.go

```diff
@@ -153,60 +153,71 @@ func (r Run) Run(ctx context.Context) error {
 	waiter.Wait()
 	<-errDoneCh
 
-	return multierror.Append(merrorWaiter, scanner.Err()).ErrorOrNil()
+	if reader.Err() != nil {
+		printError(commandFromContext(r.c), r.c.Command.Name, reader.Err())
+	}
+
+	return multierror.Append(merrorWaiter, reader.Err()).ErrorOrNil()
 }
 
-// Scanner is a cancelable scanner.
-type Scanner struct {
-	*bufio.Scanner
+// Reader is a cancelable reader.
+type Reader struct {
+	*bufio.Reader
 	err    error
 	linech chan string
 	ctx    context.Context
 }
 
-// NewScanner creates a new scanner with cancellation.
-func NewScanner(ctx context.Context, r io.Reader) *Scanner {
-	scanner := &Scanner{
-		ctx:     ctx,
-		Scanner: bufio.NewScanner(r),
-		linech:  make(chan string),
+// NewReader creates a new reader with cancellation.
+func NewReader(ctx context.Context, r io.Reader) *Reader {
+	reader := &Reader{
+		ctx:    ctx,
+		Reader: bufio.NewReader(r),
+		linech: make(chan string),
 	}
 
-	go scanner.scan()
-	return scanner
+	go reader.read()
+	return reader
 }
 
-// scan read the underlying reader.
-func (s *Scanner) scan() {
-	defer close(s.linech)
+// read reads lines from the underlying reader.
+func (r *Reader) read() {
+	defer close(r.linech)
 
 	for {
 		select {
-		case <-s.ctx.Done():
-			s.err = s.ctx.Err()
+		case <-r.ctx.Done():
+			r.err = r.ctx.Err()
 			return
 		default:
-			if !s.Scanner.Scan() {
-				return
+			// If ReadString encounters an error before finding a delimiter,
+			// it returns the data read before the error and the error itself (often io.EOF).
+			line, err := r.ReadString('\n')
+			if line != "" {
+				r.linech <- line
+			}
+			if err != nil {
+				if err == io.EOF {
+					return
+				}
+				r.err = multierror.Append(r.err, err)
 			}
-
-			s.linech <- s.Scanner.Text()
 		}
 	}
 }
 
-// Scan returns read-only channel to consume lines.
-func (s *Scanner) Scan() <-chan string {
-	return s.linech
+// Read returns read-only channel to consume lines.
+func (r *Reader) Read() <-chan string {
+	return r.linech
 }
 
 // Err returns encountered errors, if any.
-func (s *Scanner) Err() error {
-	if s.err != nil {
-		return s.err
+func (r *Reader) Err() error {
+	if errors.Is(r.err, context.Canceled) {
```

Comment:
I think we can return `r.err` directly without checking the type of the error.

Reply:
done